### ISSUE TITLE
En-Passant

### ISSUE THREAD

**Issue by EarlRussellJr** [oakmac/chessboardjs#134]
Hello, I am just using Chessboard.js in my app and when playing an existing game when the en passant move occurs the correct movement of the pawn occurs but the captured pawn is not removed.  Below is my code and the game.  Can you tell me how to enter this chess move that makes the capture and removes the pawn?  Or have I missed something in the cfg method?

On another note, I would request you add information to your documenation as to how the developer should program the board for a castle (you provided me the answer to this last week in another issue), en passant and a pawn promotion.  There is lots to read but there is no coverage of these specific topics.  I know your product is a basic board with no rules, but it would go a long way in helping others to program your product.

Thank you,
Earl Russell

```
/* Chessboard */
var board;

var onSnapEnd = function () {
    board.position(game.fen());
};

var cfg = {
    position: 'start',
    showNotation: false,
    onSnapEnd: onSnapEnd
};
```

 $(function ()
        board = ChessBoard('board', cfg);

Follow this game and on move 29 white captures a black pawn en passant.  The White pawn moves correctly but the black captured pawn is not removed from the board.

What is missing that removes the captured pawn???

1.e4 e5 2.Nf3 Nc6 3.Bb5 a6 4.Ba4 Nf6 5.O-O Nxe4 6.d4 b5 7.Bb3 d5 8.dxe5 Be6
9.Nbd2 Nc5 10.c3 Be7 11.Bc2 d4 12.cxd4 Nxd4 13.Nxd4 Qxd4 14.Nf3 Qxd1 15.Rxd1 Bg4
16.Be3 O-O 17.h3 Bxf3 18.gxf3 Rfd8 19.f4 g6 20.Kg2 Na4 21.Rxd8+ Rxd8 22.Bxa4 bxa4
23.Rc1 c5 24.Rc4 a3 25.b3 Kf8 26.Kf3 Ke8 27.Ke4 Rd1 28.Rc1 f5+ 29.exf6 Rxc1
30.Bxc1 Bxf6 31.Bxa3 Bd4 32.f3 Kd7 33.Kd5 Be3 34.f5 gxf5 35.Bxc5 Bf4 36.b4 Bd2
37.Bd6 Bc1 38.a4 Ba3 39.Bc5 Bb2 40.Bd4 Ba3 41.b5 axb5 42.axb5 Be7 43.b6  1-0
1. e2-e4 e7-e5 2. g1-f3 b8-c6 3. f1-b5 a7-a6 4. b5-a4 g8-f6 5. e1-g1,h1-f1 f6-e4 6. d2-d4 b7-b5 7. a4-b3 d7-d5 8. d4-e5 c8-e6 9. b1-d2 e4-c5 10. c2-c3 f8-e7 11. b3-c2 d5-d4 12. c3-d4 c6-d4 13. f3-d4 d8-d4 14. d2-f3 d4-d1 15. f1-d1 e6-g4 16. c1-e3 e8-g8, h8-f8 17. h2-h3 g4-f3 18. g2-f3 f8-d8 19. f3-f4 g7-g6 20. g1-g2 c5-a4 21. d1-d8 a8-d8 22. c2-a4 b5-a4 23. a1-c1 c7-c5 24. c1-c4 a4-a3 25. b2-b3 g8-f8 26. g2-f3 f8-e8 27. f3-e4 d8-d1 28. c4-c1 f7-f5 29. e5-f6 d1-c1 30. e3-c1 e7-f6 31. c1-a3 f6-d4 32. f2-f3 e8-d7 33. e4-d5 d4-e3 34. f4-f5 g6-f5 35. a3-c5 e3-f4 36. b3-b4 f4-d2 37. c5-d6 d2-c1 38. a2-a4 c1-a3 39. d6-c5 a3-b2 40. c5-d4 b2-a3 41. b4-b5 a6-b5 42. a4-b5 a3-e7 43. b5-b6

**Comment by user14159** [comment]
It works fine over here. Are you using chess.js for move validation and piece movement? 

https://github.com/jhlywa/chess.js/blob/master/README.md

**Comment by EarlRussellJr** [comment]
I have followed examples using chess.js and it does not work.  I have used
their code that works on the demo page but does not work on my code.
Haven't figured out why my code does not work.

Not sure what you mean by 'works fine over here'.  What I would like to
know, when writing my own code, is how to make the pawn move and at the
same time remove the en passent pawn in one movement just using my code and
chessboard.js.

Your help and suggestions are appreciated.

Thank you

**Comment by user14159** [comment]
Just saying the en passant functions correctly on my installation. Could you post your entire code?

**Comment by EarlRussellJr** [comment]
I have one page of code that does not use chess.js and another page that
uses it.  I am more interested in understanding how to make the move to
chessboard.js.

For example, you pass two move scripts for a castle operation.  And your
code handles a promotion.  Excluding the logic on determining an en passant
condition I haven't not seen or figured out how to tell chessboard.js to
remove the captured pawn.

Once I understand how chessboard.js handles the call then I can complete my
code and figure out the issue with my implementation of chess.js.

You can see my page and code at http://rsichess.russellsoft.com/.  This is
a test site and I ask that it not be shared or advertised at this time.
The code using chess.js is not installed at this time.

Thank you.

**Comment by user14159** [comment]
Ah, the problem is in replaying a game and not in playing a live game?

**Comment by EarlRussellJr** [comment]
Actually the problem is with both.  It was first discovered in playing a
game that actually worked correctly but never removed the pawn.  So I
assumed, right or wrong, that the chessboard.js code knew the pawn should
not be there but it was not removed.  I then assumed the move was missing
something in the script (based on how the castle is done).

I am working on playing a live game using chess.js and again ran into the
problem.

So how does chessboard.js know to remove the en passant pawn?

Thank you.

**Comment by user14159** [comment]
Well, this isn't my code so I don't know it in and out. But chessboard.js gets the en passant and castling logic from chess.js. The top of your page should contain:

```
  <script src="/js/jquery-1.10.1.min.js"></script>
  <script src="/js/chessboard-0.3.0.min.js"></script>
  <script src="/js/chess.min.js"></script>
```

You could check the game's FEN after an en passant to see if that's correct or if it's a graphical problem. Also whether the pawn remains there for just one move and disappears when the board is redrawn after the next move, or it stays there for the duration of the game.

**Comment by EarlRussellJr** [comment]
Interesting.  I will recheck to make sure chess.js is loaded and in memory
before starting and then retest.  I did not know that chessboard.js
depended upon chess.js code.

In the cases that I have tested the pawn was never removed.  I have never
checked the FEN except when I want to capture a position.  I intend to use
this more when I tackle the 'move backward' feature.

But I still do not understand how chessboard.js is to be instructed to
remove an en passant pawn.  Can you share that information?

Thanks again for your assistance.

**Comment by EarlRussellJr** [comment]
Correction - opening the first page's link (Openings) I do have chess.js
included.  Likewise I have it for the Archived pages.  So I do not think
this is the solution as the problem occurs with both libraries in memory.

QUESTION: Does the order of .js files loaded have any importance?
<title>Chess Openings</title>
<link href="/rsichess/Content/ChessAnalytics.css
<http://rsichess.russellsoft.com/rsichess/Content/ChessAnalytics.css>" rel="
stylesheet" />
<link href="/rsichess/Content/chessboard-0.3.0.css
<http://rsichess.russellsoft.com/rsichess/Content/chessboard-0.3.0.css>" rel
="stylesheet" />
<script src="/rsichess/Scripts/chessboard-0.3.0.js
<http://rsichess.russellsoft.com/rsichess/Scripts/chessboard-0.3.0.js>">
</script>
<script src="/rsichess/Scripts/chess.js
<http://rsichess.russellsoft.com/rsichess/Scripts/chess.js>"></script>
<script src="/rsichess/Scripts/ChessAnalytics.js
<http://rsichess.russellsoft.com/rsichess/Scripts/ChessAnalytics.js>">
</script>

**Comment by user14159** [comment]
Just took a look at your site and it does look like you're correctly loading chess.js. 

Removing an en passant pawn is going to be similar to a normal capture. However, in the case of a normal capture, a piece is replacing the location of the captured piece. You can see an example of how a piece is removed from the board without being replaced by another in the example here: [http://chessboardjs.com/examples#2082]. When an en passant occurs, the new legal position of all the pieces will be determined by chess.js. The board is then redrawn at the end of each move with the pieces in the same positions as indicated by the FEN. Take a look at this function:

```
   var onSnapEnd = function() {
    board.position(game.fen());
   };

```

**Comment by EarlRussellJr** [comment]
Great!  I will get back to work and follow your instructions.

Thank you.

**Comment by user14159** [comment]
The ordering does make a difference. For example jquery has to come first. But in your case it looks okay. Could you give some game on your site featuring this non-disappearing pawn problem?

**Comment by EarlRussellJr** [comment]
Give me a few minutes - have to find it but I think it is an Alekine game.

**Comment by EarlRussellJr** [comment]
Choose Archived Games
Select Grand Master Archives
Select Alekhine
Select Draws
Load Games
Enter 282 and Go To game
1932 Alekhine vs Fine
Move 30 is where it happens - follow afterwards and a rook move goes right
thru the unmoved pawn, so Chessboard.js knows the pawn is gone but it was
not removed from the board.

I played this game in my local code, made sure chess.js was loaded before
chessboard.js, and the 'trash' like was added to the cfg method.  The pawn
was not removed.

**Comment by oakmac** [comment]
> When an en passant occurs, the new legal position of all the pieces will be determined by chess.js. The board is then redrawn at the end of each move with the pieces in the same positions as indicated by the FEN.

@user14159 is correct here. chessboard.js has no knowledge of en passant. You need to update the board state from the position provided by chess.js after the en passant capture.

**Comment by EarlRussellJr** [comment]
OK. I make the assumption that chess.js computes the piece positions after
an en passant action and then resets chessboard.js using the computed FEN
before the next move occurs.  Is that correct?

Based on your input I will have to implement chess.js into my pages and
rewrite some of the code to use chess.js.  I was hoping not to have to do
this but I suppose it is easier rather than write my own methods to
determine when and if an en passant action occurs.

However, I do have a live game page working so I can add the FEN action to
that page and if all works fine then use that code for the other pages.

*JUST ONE MORE QUESTION - an easy one for you...*
Is there any way to change the board colors?
I have a few players testing my site and one had asked to change the
colors.  Personally I like the colors, but some others would like the
option.

Many thanks for your help.

**Comment by oakmac** [comment]
Most chessboard colors can be changed using CSS.

Check out: https://github.com/oakmac/chessboardjs/blob/master/www/css/chessboard.css

**Comment by oakmac** [comment]
> chess.js computes the piece positions after an en passant action

Yes

> and then resets chessboard.js using the computed FEN before the next move occurs. Is that correct?

Yes

**Comment by EarlRussellJr** [comment]
Thank you.  It's been great chatting with you.

**Comment by user14159** [comment]
Had some more time to look over your code and I'm guessing the problem is in your `findMove()` function. Look over the details of FEN (https://en.wikipedia.org/wiki/Forsyth%E2%80%93Edwards_Notation). The FEN code itself contains information on which side can castle K-side or Q-side and also en passant details. Unless I'm misunderstanding, it looks like you're rewriting this code yourself based off the PGN. The legal moves should already be defined in chess.js, represented by the FEN, and then drawn onto the board with chessboard.js, just as you said.

**Comment by EarlRussellJr** [comment]
Thank you for your feedback - you are spot on.  I wrote the original code
for Openings to play from the play list and did all my own code to handle
all the interactions between the move list, manually moving, auto moving,
game selection and so on - all to interact with chessboard.js.  I did not
use chess.js at all.  I wrote my own code to determine castling and you
already had promotion so that worked.  Then I found a game with en passant
and hit the wall.

Not sure I can fix this in my code and may have to rewrite to use chess.js
in place of my code.  At least I now know what to do to fix the issue.

Thanks!
